### PR TITLE
Add support for DS1307 real time clock

### PR DESCRIPTION
**Pull request by badbadc0ffee**
## Description:

**Pull request in [ESPHome](https://github.com/esphome/ESPHome) with YAML changes (if applicable):** esphome/ESPHome#399
**Pull request in [ESPHome-Docs](https://github.com/esphome/ESPHome-Docs) with C++ framework changes (if applicable):** esphome/ESPHome-Docs#144

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphomedocs](https://github.com/OttoWinter/esphomedocs).

### CODE REVIEW

File: src/esphomelib/time/ds1307.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+void DS1307Time::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up DS1307...");
+  if (!read_data()) {
+    mark_failed();
+  }
+}
+
+bool DS1307Time::read_data() {
+  if (!read_bytes(0, ds1307_.raw, sizeof(ds1307_.raw))) {
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]:
```suggestion
  if (!this->read_bytes(0, this->ds1307_.raw, sizeof(this->ds1307_.raw))) {
```

File: src/esphomelib/time/ds1307.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  time.strftime(buf, sizeof(buf), "%c");
+  ESP_LOGD(TAG, "Synchronized time: %s", buf);
+  ESP_LOGV(TAG, "Rate select: %0d", ds1307_.reg.rs);
+  ESP_LOGV(TAG, "Square-Wave Enable (SQWE): %0d", ds1307_.reg.sqwe);
+  ESP_LOGV(TAG, "Output Control (OUT): %0d", ds1307_.reg.out);
+  return true;
+}
+
+bool DS1307Time::write_data() {
+  if (!read_bytes(0, ds1307_.raw, sizeof(ds1307_.raw))) {
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]:
```suggestion
  if (!this->read_bytes(0, this->ds1307_.raw, sizeof(this->ds1307_.raw))) {
```

File: src/esphomelib/time/ds1307.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    ESP_LOGE(TAG, "Can't read I2C data.");
+    return false;
+  }
+  time::EsphomelibTime time = now();
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]:
```suggestion
  time::EsphomelibTime time = this->now();
```

File: src/esphomelib/time/ds1307.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  ds1307_.reg.day       = time.day_of_month % 10;
+  ds1307_.reg.day_10    = time.day_of_month / 10;
+  ds1307_.reg.hour      = time.hour         % 10;
+  ds1307_.reg.hour_10   = time.hour         / 10;
+  ds1307_.reg.minute    = time.minute       % 10;
+  ds1307_.reg.minute_10 = time.minute       / 10;
+  ds1307_.reg.second    = time.second       % 10;
+  ds1307_.reg.second_10 = time.second       / 10;
+  ds1307_.reg.weekday   = time.day_of_week;
+  if (!write_bytes(0, ds1307_.raw, sizeof(ds1307_.raw))) {
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]:
```suggestion
  if (!this->write_bytes(0, this->ds1307_.raw, sizeof(this->ds1307_.raw))) {
```

File: src/esphomelib/time/ds1307.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    ESP_LOGE(TAG, "Can't write I2C data.");
+    return false;
+  }
+  char buf[128];
+  time.strftime(buf, sizeof(buf), "%c");
+  ESP_LOGD(TAG, "Synchronized time: %s", buf);
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]: If this integration both writes + reads time, there should be a difference in log output between the two.

File: src/esphomelib/time/ds1307.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+class DS1307Time : public RealTimeClockComponent, public I2CDevice {
+ public:
+  DS1307Time(I2CComponent *parent, uint8_t address);
+  void setup() override;
+  void dump_config() override;
+  float get_setup_priority() const override;
+  bool read_data();
+  bool write_data();
+ private:
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.h]:
```suggestion
 protected:
```

File: src/esphomelib/time/ds1307.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  typedef union {
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.h]:
```suggestion
  union DS1307Reg {
```

**Review by OttoWinter — COMMENTED** [review]
Thanks!

Left you some comments (mostly style)

Please also update defines.h in `ESPHOMEYAML_USE`

File: src/esphomelib/time/ds1307.cpp (hunk 3 above)
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]: `now()` returns the localized time. Shouldn't the UTC time be written to the clock?

At least `settimeofday` expects UTC time values AFAIK, so if you would call read/write repeatedly when in a non-UTC timezone, you would increment the hour each time.

Comment (badbadc0ffee) [reply src/esphomelib/time/ds1307.cpp]: That needs to be checked indeed! Right now, I am in the Germany/Berlin TZ and the time appears not to be incrementing. But you are right; it should...

Comment (badbadc0ffee) [reply src/esphomelib/time/ds1307.cpp]: This is odd. You should be right. I should write UTC to the H/W clock, but when I do that, the time is st back one hour every time I write and read back from the H/W clock. Will investigate further on this...

File: src/esphomelib/time/ds1307.cpp [hunk 8]
```diff
@@ -52,8 +52,7 @@ bool DS1307Time::read_data() {
   t.tm_min  = ds1307_.reg.minute + 10 * ds1307_.reg.minute_10;
   t.tm_sec  = ds1307_.reg.second + 10 * ds1307_.reg.second_10;
   struct timeval tv {::mktime(&t), 0};
-  timezone tz = {0, 0};
-  ::settimeofday(&tv, &tz);
+  ::settimeofday(&tv, 0);
```
Comment (OttoWinter) [inline src/esphomelib/time/ds1307.cpp]: Revert this please. If you do not specify the second argument for `settimeofday`, the timezone will be +9h china time. (The SDK is not fully POSIX compliant, took me a while to figure that out for home assistant time.) - timezones are handled by the layer above settimeofday in the TZ env-variable.

**Comment by legomind** [comment]
Is there any movement on this? Looking to use this for an upcoming irrigation controller project.

**Comment by ripvega** [comment]
I'd really love to see this support implemented. I don't have the skills to offer much in the way of support though.

I have two ESPHome nodes, one for heating control and one that controls my Koi Pond, that both depend on time. I've made them Home Assistant independent, apart from front end variable control, but when our unreliable internet fails, it causes real headaches.

**Comment by albasee** [comment]
I also really would love to see this support implemented. Thanks in advance

**Comment by badbadc0ffee** [comment]
will be replaced by another PR for the "new" repository layout

**Comment by badbadc0ffee** [comment]
check out https://github.com/esphome/esphome/pull/1441

**Pull request closed without merging**